### PR TITLE
CSI: remove prefix matching from CSIVolumeByID and fix CLI prefix matching

### PR DESCRIPTION
**Pull request by tgross**
Callers of `CSIVolumeByID` are generally assuming they should receive a single volume. Among other possible bugs, this potentially results in feasibility checking being performed against the wrong volume if a volume's ID is a prefix substring of other volume. Discovered this while working on https://github.com/hashicorp/nomad/pull/10136 because of `"test"` vs `"test[0]"`

But removing the incorrect prefix matching from `CSIVolumeByID` breaks prefix matching in the command line client. Add the required elements for prefix matching back to the commands and API.

### CODE REVIEW

File: api/csi.go [hunk 1]
```diff
@@ -28,11 +28,6 @@ func (v *CSIVolumes) List(q *QueryOptions) ([]*CSIVolumeListStub, *QueryMeta, er
 	return resp, qm, nil
 }
 
-// PluginList returns all CSI volumes for the specified plugin id
-func (v *CSIVolumes) PluginList(pluginID string) ([]*CSIVolumeListStub, *QueryMeta, error) {
-	return v.List(&QueryOptions{Prefix: pluginID})
```
Comment (tgross) [inline api/csi.go:33]: `CSIVolumes.List` has always been how you get this info, and this code has never worked. It has no callers either, so I'm removing here... I can't imagine this will break anyone's code given that it didn't work.

File: nomad/csi_endpoint.go [hunk 2]
```diff
@@ -120,15 +120,19 @@ func (v *CSIVolume) List(args *structs.CSIVolumeListRequest, reply *structs.CSIV
 			if err != nil {
 				return err
 			}
+			// TODO: handle prefix arg
+
```
Comment (cgbaker) [inline nomad/csi_endpoint.go:124]:
```suggestion
```

Comment (cgbaker) [reply nomad/csi_endpoint.go:124]: assuming this is sufficiently handled below?

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -2126,26 +2126,26 @@ func (s *StateStore) CSIVolumes(ws memdb.WatchSet) (memdb.ResultIterator, error)
 func (s *StateStore) CSIVolumeByID(ws memdb.WatchSet, namespace, id string) (*structs.CSIVolume, error) {
 	txn := s.db.ReadTxn()
 
-	watchCh, obj, err := txn.FirstWatch("csi_volumes", "id_prefix", namespace, id)
+	obj, err := txn.First("csi_volumes", "id", namespace, id)
```
Comment (cgbaker) [inline nomad/state/state_store.go:2129]: why the change from `FirstWatch` to `First`? won't this break blocking queries against `CSIVolume.Get`:
https://github.com/hashicorp/nomad/blob/9138d374d62a222159135a0b06ff91e3b134ea57/nomad/csi_endpoint.go#L208

File: nomad/state/state_store.go [hunk 4]
```diff
@@ -2212,11 +2212,13 @@ func (s *StateStore) CSIVolumesByNodeID(ws memdb.WatchSet, nodeID string) (memdb
 	iter := NewSliceIterator()
 	txn := s.db.ReadTxn()
 	for id, namespace := range ids {
-		raw, err := txn.First("csi_volumes", "id", namespace, id)
-		if err != nil {
-			return nil, fmt.Errorf("volume lookup failed: %s %v", id, err)
+		if strings.HasPrefix(id, prefix) {
+			raw, err := txn.First("csi_volumes", "id", namespace, id)
```
Comment (cgbaker) [inline nomad/state/state_store.go:2216]: this isn't a change to the previous behavior, but i wonder whether this should also be a `FirstWatch`, so that a blocking query against this endpoint will return if one of the volumes changes. 
this may have been intended by the call here: 
https://github.com/hashicorp/nomad/pull/10158/files#diff-53c7d524897a192ece819816a9dcf006dee46747fbb853e7f638601ea556a72cR2224
but unfortunately `WatchCh()` returns `nil` for `SliceIterator` objects 😢

File: command/volume_deregister.go [hunk 5]
```diff
@@ -91,6 +93,24 @@ func (c *VolumeDeregisterCommand) Run(args []string) int {
 		return 1
 	}
 
+	// Prefix search for the volume
```
Comment (cgbaker) [inline command/volume_deregister.go:96]: 🏆  these are so helpful to have...

File: command/volume_deregister.go [hunk 6]
```diff
@@ -91,6 +93,24 @@ func (c *VolumeDeregisterCommand) Run(args []string) int {
...
+		sort.Slice(vols, func(i, j int) bool { return vols[i].ID < vols[j].ID })
+		out, err := csiFormatSortedVolumes(vols, shortId)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Error formatting: %s", err))
+			return 1
+		}
+		c.Ui.Error(fmt.Sprintf("Prefix matched multiple volumes\n\n%s", out))
+		return 1
+	}
+	volID = vols[0].ID
```
Comment (cgbaker) [inline command/volume_deregister.go:112]: what if `len(vols) == 0` (no match)? see, e.g.:
https://github.com/hashicorp/nomad/blob/9138d374d62a222159135a0b06ff91e3b134ea57/command/job_stop.go#L134-L137

File: command/volume_detach.go [hunk 7]
```diff
@@ -112,6 +114,25 @@ func (c *VolumeDetachCommand) Run(args []string) int {
...
+			sort.Slice(vols, func(i, j int) bool { return vols[i].ID < vols[j].ID })
+			out, err := csiFormatSortedVolumes(vols, shortId)
+			if err != nil {
+				c.Ui.Error(fmt.Sprintf("Error formatting: %s", err))
+				return 1
+			}
+			c.Ui.Error(fmt.Sprintf("Prefix matched multiple volumes\n\n%s", out))
+			return 1
+		}
+		volID = vols[0].ID
```
Comment (cgbaker) [inline command/volume_detach.go:134]: similar to deregister; need to handle `len(vols) == 0`

File: command/volume_status_csi.go [hunk 8]
```diff
@@ -39,6 +39,23 @@ func (c *VolumeStatusCommand) csiStatus(client *api.Client, id string) int {
...
+	if len(vols) > 1 {
+		out, err := c.csiFormatVolumes(vols)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Error formatting: %s", err))
+			return 1
+		}
+		c.Ui.Error(fmt.Sprintf("Prefix matched multiple volumes\n\n%s", out))
+		return 1
+	}
+	id = vols[0].ID
```
Comment (cgbaker) [inline command/volume_status_csi.go:57]: and one more `len(vols) == 0` case, i think

**Review by cgbaker — APPROVED** [review]
a couple of questions about blocking queries and CLI handling when prefix search returns no matches.

File: nomad/state/state_store.go (hunk 3 above)
Comment (tgross) [reply nomad/state/state_store.go:2129]: Good catch.

File: nomad/csi_endpoint.go (hunk 2 above)
Comment (tgross) [reply nomad/csi_endpoint.go:124]: Whoops! Note left to myself there.

File: nomad/state/state_store.go (hunk 3 above)
Comment (tgross) [reply nomad/state/state_store.go]: Fixed in 5836f3b

File: nomad/state/state_store.go (hunk 4 above)
Comment (tgross) [reply nomad/state/state_store.go]: Agreed. Fixed in 5836f3b
